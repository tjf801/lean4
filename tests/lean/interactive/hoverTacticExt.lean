


/-!
Tests that docstring hovers are computed correctly for tactic extensions
-/

/-- Another `trivial` tactic -/
syntax (name := my_trivial) "my_trivial" : tactic

@[tactic_alt my_trivial]
syntax (name := very_trivial) "very_trivial" : tactic
macro_rules
| `(tactic|very_trivial) => `(tactic|my_trivial)

/-- It tries Lean's `trivial` -/
tactic_extension my_trivial
               --^ textDocument/hover
macro_rules
  | `(tactic|my_trivial) => `(tactic|trivial)
             --^ textDocument/hover

/-- It tries `simp_all` -/
tactic_extension my_trivial
macro_rules
  | `(tactic|my_trivial) => `(tactic|simp_all; done)


example : True := by
  my_trivial
      --^ textDocument/hover

/-- It tries `constructor` -/
tactic_extension my_trivial
               --^ textDocument/hover
-- On the preceding line, it was not yet extended.
-- Here, it is:
macro_rules
  | `(tactic|my_trivial) => `(tactic|constructor; done)
           --^ textDocument/hover

/--
It tries some useful things:
 * `omega`
 * `simp_arith [*]`
-/
tactic_extension my_trivial
macro_rules
  | `(tactic|my_trivial) => `(tactic|omega)
macro_rules
  | `(tactic|my_trivial) => `(tactic|simp_arith [*]; done)

-- This tests that nested lists work
example : True := by
  my_trivial
      --^ textDocument/hover

-- This tests that alts are resolved
example : True := by
  very_trivial
      --^ textDocument/hover
