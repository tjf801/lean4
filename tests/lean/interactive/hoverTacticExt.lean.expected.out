{"textDocument": {"uri": "file:///hoverTacticExt.lean"},
 "position": {"line": 16, "character": 17}}
{"range":
 {"start": {"line": 16, "character": 17}, "end": {"line": 16, "character": 27}},
 "contents":
 {"value":
  "```lean\nmy_trivial : Lean.ParserDescr\n```\n***\nAnother `trivial` tactic ",
  "kind": "markdown"}}
{"textDocument": {"uri": "file:///hoverTacticExt.lean"},
 "position": {"line": 19, "character": 15}}
{"range":
 {"start": {"line": 19, "character": 13}, "end": {"line": 19, "character": 23}},
 "contents":
 {"value":
  "Another `trivial` tactic \n\nExtensions:\n\n * It tries Lean's `trivial`",
  "kind": "markdown"}}
{"textDocument": {"uri": "file:///hoverTacticExt.lean"},
 "position": {"line": 29, "character": 8}}
{"range":
 {"start": {"line": 29, "character": 2}, "end": {"line": 29, "character": 12}},
 "contents":
 {"value":
  "Another `trivial` tactic \n\nExtensions:\n\n * It tries Lean's `trivial` \n\n * It tries `simp_all`",
  "kind": "markdown"}}
{"textDocument": {"uri": "file:///hoverTacticExt.lean"},
 "position": {"line": 33, "character": 17}}
{"range":
 {"start": {"line": 33, "character": 17}, "end": {"line": 33, "character": 27}},
 "contents":
 {"value":
  "```lean\nmy_trivial : Lean.ParserDescr\n```\n***\nAnother `trivial` tactic \n\nExtensions:\n\n * It tries Lean's `trivial` \n\n * It tries `simp_all`",
  "kind": "markdown"}}
{"textDocument": {"uri": "file:///hoverTacticExt.lean"},
 "position": {"line": 38, "character": 13}}
{"range":
 {"start": {"line": 38, "character": 13}, "end": {"line": 38, "character": 23}},
 "contents":
 {"value":
  "Another `trivial` tactic \n\nExtensions:\n\n * It tries Lean's `trivial` \n\n * It tries `simp_all` \n\n * It tries `constructor`",
  "kind": "markdown"}}
{"textDocument": {"uri": "file:///hoverTacticExt.lean"},
 "position": {"line": 54, "character": 8}}
{"range":
 {"start": {"line": 54, "character": 2}, "end": {"line": 54, "character": 12}},
 "contents":
 {"value":
  "Another `trivial` tactic \n\nExtensions:\n\n * It tries Lean's `trivial` \n\n * It tries `simp_all` \n\n * It tries `constructor` \n\n * It tries some useful things:\n    * `omega`\n    * `simp_arith [*]`",
  "kind": "markdown"}}
{"textDocument": {"uri": "file:///hoverTacticExt.lean"},
 "position": {"line": 59, "character": 8}}
{"range":
 {"start": {"line": 59, "character": 2}, "end": {"line": 59, "character": 14}},
 "contents":
 {"value":
  "Another `trivial` tactic \n\nExtensions:\n\n * It tries Lean's `trivial` \n\n * It tries `simp_all` \n\n * It tries `constructor` \n\n * It tries some useful things:\n    * `omega`\n    * `simp_arith [*]`",
  "kind": "markdown"}}
